### ISSUE TITLE
Cryptofuzz in OpenSSL CI

### ISSUE THREAD

**Issue by guidovranken** [openssl/openssl#9744]
As the master branch keeps receiving many updates, new bugs keep getting found in Cryptofuzz/OSS-Fuzz.

I think the quality of your software could benefit from a CI integration of Cryptofuzz. Cryptofuzz + a static corpus (eg. never or infrequently updated) and without differential capabilities (eg. don't test against other libraries) can detect many bugs as soon as they are attempted to be merged.

Just today 2 new bugs were found:

- A UAF https://oss-fuzz.com/testcase-detail/5658978675064832
- Invalid read https://oss-fuzz.com/testcase-detail/5696325059870720

(might be related to one another)

These are probably due to some recent change, and if Cryptofuzz was part of your CI tests, you might have been alerted sooner.

How do you feel about this?

**Comment by mattcaswell** [comment]
I think this would fit really well. We already have the concept of "external" tests (e.g. we run the boringssl tests, pyca tests and krb5 external tests suites)...although they need some cleaning up because the pyca tests have been failing for a long while now.

**Comment by levitte** [comment]
Not sure if I understand this correctly, but when seeing someone talk about CIs, I imagine something hooked in via github, alongside Travis and Appveyor.  Is that what's suggested?

**Comment by guidovranken** [comment]
Yes, run alongside Travis and AppVeyor. My idea is to run the following upon each pull request, or change to an existing PR:

* Compile OpenSSL (with the PR integrated) with AddressSanitizer (+ maybe ubsan).

* Compile Cryptofuzz (pinned to a specific commit that is known to work to ensure I don't break anything while working on Cryptofuzz).

* Get a recent corpus, that I can provide in https://github.com/guidovranken/cryptofuzz-corpora

* Run ```./cryptofuzz corpus/ -runs=1```

If this triggers AddressSanitizer (we can tell from the exit code) or an abort, there's a bug.

**Comment by levitte** [comment]
So how do we hook that in?

**Comment by guidovranken** [comment]
I can write a bash script that performs all the stepts I listed, which I suppose you can call from Travis  -- or should I edit travis.yml directly?

**Comment by mattcaswell** [comment]
My preferred approach would be to do this in the same way that we have done for pyca (etc) external tests. So create an external test `test/recipes/95-test_external_cryptofuzz.t` modelled on the pyca equivalent. That test also has a shell script that sets everything up (in `test/recipes/95-test_external_pyca_data/cryptography.sh`). We can have cryptofuzz as a submodule under the top level directory (as we do for pyca/boringssl/krb5) associated with a particular "known good" commit.

Once all that is done it is a simple matter of ensure that there is a travis job that runs the external test.

**Comment by levitte** [comment]
There is a Travis job that runs all external tests.

Sorry about earlier, I had in my mind that cryptofuzz was a separate CI service like Travis. 

I do worry about having cryptofuzz as a submodule. How much data are we talking about, if someone (like me) wants to run the external tests on their laptop, say?

**Comment by levitte** [comment]
For Travis, feel free. You can probably see how we did things for other external stuff, most prominently BoringSSL. Either way, you will need a script in test/recipes that drives a cryptofuzz test, and you can add whatever condition needed to run that test. Previous ones can be used as examples, they all have "external" as part of their file names. 

I think 12 MB is acceptable for this, as long as it's an optional thing (I.e the test script can detect its presence, and simply skip if not present)

**Comment by guidovranken** [comment]
I mean once ```test/recipes/95-test_external_cryptofuzz_data/cryptofuzz.sh``` is entered, OpenSSL has already been compiled, correct? Then how do I ensure that OpenSSL gets compiled with the correct flags?

**Comment by levitte** [comment]
That, you do with configuration flags. In our Travis file, we do that through a per job environment variable, CONFIG_OPTS (I hope I spelt it right). You can use other environment variables for further control as needed. Have a look at what's already there...

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
